# Atividade Prática — Aula 4: Análise de Dados e Construção de Indicadores (KPIs)

Esta atividade foi construída com base na Aula 4, que marca a transição **do dado limpo para a tomada de decisão estratégica**, com foco em **KPIs, métricas derivadas, dimensões de análise, rankings e insights executivos**. fileciteturn5file0

## Ideia central da aula
Antes do código, vem a **pergunta de negócio**.  
Depois, definimos:
1. a **métrica**
2. a **dimensão**
3. a **agregação**
4. e só então construímos a resposta em tabela e em texto. fileciteturn5file0

## Regras da atividade
- O notebook **orienta**, mas você deve **escrever os códigos**.
- Sempre que gerar uma tabela agregada, escreva **1 ou 2 frases interpretando o resultado**.
- O objetivo não é apenas calcular números, mas comunicar decisões.

## Dataset da atividade
Arquivo: `vendas_brasil_aula4_kpis.csv`


## 1. Preparação do ambiente

Importe as bibliotecas necessárias para a atividade.

**Sugestão:**  
- `pandas`
- `numpy`

Se desejar, você também pode usar uma biblioteca de visualização mais adiante, mas o foco desta aula é **tabela agregada + insight**.


In [1]:
import pandas as pd
import numpy as np

## 2. Leitura da base

Leia o arquivo `vendas_brasil_aula4_kpis.csv` em um DataFrame chamado `df`.

Depois:
1. exiba as primeiras linhas
2. veja o tamanho da base
3. inspecione os tipos das colunas
4. confirme se `data_venda` está pronta para análises temporais


In [6]:
df = pd.read_csv('vendas_brasil_aula4_kpis.csv')
df.head()

,pedido_id,data_venda,uf,canal_venda,categoria,produto,quantidade_pedidos,preco_unitario,receita,lucro
0,5000,2025-02-14,RJ,Marketplace,Periféricos,Mouse Gamer,2,237.65,475.30,239.08
1,5001,2025-02-17,PR,Online,Informática,Monitor 27,4,1431.79,5727.16,1769.48
2,5002,2025-07-22,RJ,Marketplace,Informática,Monitor 27,5,1446.22,7231.10,2449.85
3,5003,2025-12-08,SP,Marketplace,Periféricos,Teclado Mecânico,7,399.84,2798.88,1356.67
4,5004,2025-01-04,GO,Marketplace,Telefonia,Smartphone X,4,3134.85,12539.40,3656.84


## 3. Perguntas antes do código

A aula enfatiza que a maior armadilha é começar com `groupby()` sem saber o que se quer responder. fileciteturn5file0

### Tarefa
Antes de programar, escreva em markdown respostas curtas para estas perguntas:

1. Qual canal de venda é o mais eficiente?
2. Qual categoria parece mais rentável?
3. Quais produtos são os campeões por lucro?
4. O negócio está crescendo ao longo do tempo?
5. Quais KPIs globais resumem melhor a operação?

Depois disso, avance para o código.


## 4. KPIs globais da operação

Com base nas “4 Métricas de Ouro do Varejo” apresentadas na aula, calcule os seguintes KPIs globais da base inteira: fileciteturn5file0

- **Receita Total**
- **Lucro Total**
- **Margem de Lucro (%)**
- **Ticket Médio**

### Orientação
- Pense primeiro na fórmula de cada KPI
- Escreva os cálculos em células separadas ou em uma pequena tabela-resumo
- Se quiser, formate os resultados para facilitar a leitura


In [8]:
# Calcule aqui os 4 KPIs globais

receita_total = df['receita'].sum()
lucro_total = df['lucro'].sum()
margem_lucro_percentual = (lucro_total / receita_total) * 100
ticket_medio = receita_total / df['pedido_id'].nunique()

print(f"Receita Total: R$ {receita_total:,.2f}")
print(f"Lucro Total: R$ {lucro_total:,.2f}")
print(f"Margem de Lucro: {margem_lucro_percentual:.2f}%")
print(f"Ticket Médio: R$ {ticket_medio:,.2f}")



Receita Total: R$ 1,921,492.22
Lucro Total: R$ 544,496.15
Margem de Lucro: 28.34%
Ticket Médio: R$ 4,574.98


### Interpretação
Escreva 3 a 5 linhas respondendo:
- O negócio parece grande ou pequeno?

o negócio parece ser de grande porte, uma vez que ao analisarmos a sua receita total, o valor de 1,921,492.22 sugere alto fluxo de caixa

- O lucro acompanha bem a receita?

O lucro aparenta estar em um valor satisfatório, visto que a receita total está proxima de dois milhões e o lucro pouco acima de meio milhão

- O ticket médio sugere compras de maior valor ou de menor valor?
o ticket medio sugere compras de maior valor, podendo ser mais de 4 mil ou menos de 4 mil


## 5. Métricas derivadas

A aula mostra que **somar valores não basta**.  
Precisamos criar métricas derivadas para avaliar eficiência e comportamento. fileciteturn5file0

### Tarefa
Crie, quando fizer sentido:
- uma coluna de **margem_lucro** (`lucro / receita`)
- uma lógica para **ticket_medio** da operação ou por grupo

### Reflexão
Explique:
1. Por que receita alta não significa automaticamente melhor desempenho?

Uma receita alta não necessáriamente indica um alto desempenho, indica somente o volume de vendas, sem revelar os custos envolvidos para gerar essa receita
2. Em que situação a margem ajuda mais do que a receita bruta?
A margem é crucial para avaliar a eficiência e a sustentabilidade de um negócio, sendo útil em comparação de produtos/serviços, precificação e estratégia


In [9]:
df['margem_lucro'] = df['lucro'] / df['receita']

# Exibir as primeiras linhas com a nova coluna para verificação
display(df.head())

,pedido_id,data_venda,uf,canal_venda,categoria,produto,quantidade_pedidos,preco_unitario,receita,lucro,margem_lucro
0,5000,2025-02-14,RJ,Marketplace,Periféricos,Mouse Gamer,2,237.65,475.30,239.08,0.503009
1,5001,2025-02-17,PR,Online,Informática,Monitor 27,4,1431.79,5727.16,1769.48,0.308963
2,5002,2025-07-22,RJ,Marketplace,Informática,Monitor 27,5,1446.22,7231.10,2449.85,0.338794
3,5003,2025-12-08,SP,Marketplace,Periféricos,Teclado Mecânico,7,399.84,2798.88,1356.67,0.484719
4,5004,2025-01-04,GO,Marketplace,Telefonia,Smartphone X,4,3134.85,12539.40,3656.84,0.291628


## 6. Fatiando a realidade: análise por dimensão

A aula mostra que um KPI isolado não decide nada.  
É preciso perguntar: **como isso varia por canal, categoria, região e tempo?** fileciteturn5file0

### Tarefa
Liste, em markdown, quais colunas do dataset representam:
- métrica
    -   `quantidade_pedidos`
    -   `preco_unitario`
    -   `receita`
    -   `lucro`
    -   `margem_lucro` (métrica derivada)
- dimensão
    -   `pedido_id`
    -   `uf`
    -   `canal_venda`
    -   `categoria`
    -   `produto`
- tempo
    -   `data_venda`


## 7. Análise por canal

### Pergunta de negócio
**Qual canal de venda é o mais eficiente?**

### Tarefa
Crie uma tabela agregada por `canal_venda` contendo, no mínimo:
- Receita Total
- Lucro Total
- Margem Média ou Margem Calculada
- Quantidade de pedidos
- Ticket Médio

### Depois responda
1. Qual canal gera mais receita?
2. Qual canal gera maior eficiência?
3. Se você fosse gestor, priorizaria volume ou margem?


### Insight obrigatório
Escreva 1 ou 2 frases interpretando sua tabela, no estilo executivo:
- não apenas “o Online foi maior”
- mas o que isso **significa para a decisão**

O canal Marketplace tem maior destaque pelo volume superior de receita e lucro, refletindo sua forte atuação no mercado. entretanto, o canal Onlixe exibe a melhor margem de lucro, indicando maior eficiência, podendo ser aplicada as demais

## 8. Análise por categoria

### Pergunta de negócio
**Qual categoria é mais lucrativa e qual parece mais eficiente?**

### Tarefa
Crie uma tabela agregada por `categoria` com:
- Receita Total
- Lucro Total
- Margem
- Ticket Médio

### Depois responda
1. A categoria de maior receita também é a de maior margem?

Não. A categoria com a maior receita é 'Telefonia' (R$ 866.292,45), mas sua margem calculada é de 25,39%. A categoria com a maior margem é 'Periféricos' (43,13%), que possui uma receita significativamente menor (R$$ 866.292,45), mas sua margem calculada é de 25,39%. A categoria com a maior margem é 'Periféricos' (43,13%), que possui uma receita significativamente menor (R$ 116.923,08).

2. Existe alguma categoria que parece vender bem, mas ser pouco eficiente?

Sim, a categoria 'Telefonia' se destaca por ter a maior receita, indicando que vende muito bem. No entanto, ela apresenta a menor margem de lucro (25,39%) entre todas as categorias, sugerindo que, apesar do alto volume de vendas, sua eficiência em gerar lucro é baixa.


## 9. Análise por região (UF)

### Pergunta de negócio
**Quais UFs concentram a operação?**

### Tarefa
Monte uma tabela por `uf` com:
- Receita Total
- Lucro Total
- Participação percentual da receita
- Ticket Médio

### Depois responda
1. Quais UFs concentram maior receita?

São Paulo, Rio de Janeiro, Pernambuco e Minas Gerais respectivamente

2. Existe concentração regional?

O maior número se concentra na região sudeste exceto por pernambuco

3. Se fosse necessário expandir ou reforçar logística, por onde você começaria?

devido a alta demanda de São paulo e Rio de Janeiro seria mais coerente reforçar a logistica nessas regiões


In [15]:

analise_uf = df.groupby('uf').agg(
    receita_total=('receita', 'sum'),
    lucro_total=('lucro', 'sum'),
    quantidade_pedidos=('pedido_id', 'nunique') # Contagem de pedidos únicos
).reset_index()

# Calcular Margem Calculada (not explicitly asked for, but useful)
analise_uf['margem_calculada'] = (analise_uf['lucro_total'] / analise_uf['receita_total']) * 100

# Calcular Ticket Médio
analise_uf['ticket_medio'] = analise_uf['receita_total'] / analise_uf['quantidade_pedidos']

# Calcular Participação percentual da receita
total_receita_global = df['receita'].sum()
analise_uf['participacao_receita_percentual'] = (analise_uf['receita_total'] / total_receita_global) * 100

# Formatar para melhor visualização
analise_uf['receita_total'] = analise_uf['receita_total'].map('R$ {:,.2f}'.format)
analise_uf['lucro_total'] = analise_uf['lucro_total'].map('R$ {:,.2f}'.format)
analise_uf['margem_calculada'] = analise_uf['margem_calculada'].map('{:,.2f}%'.format) # Optional
analise_uf['ticket_medio'] = analise_uf['ticket_medio'].map('R$ {:,.2f}'.format)
analise_uf['participacao_receita_percentual'] = analise_uf['participacao_receita_percentual'].map('{:,.2f}%'.format)

display(analise_uf.sort_values(by='receita_total', ascending=False))


,uf,receita_total,lucro_total,quantidade_pedidos,margem_calculada,ticket_medio,participacao_receita_percentual
9,SP,"R$ 246,835.50","R$ 71,022.16",48,28.77%,"R$ 5,142.41",12.85%
6,RJ,"R$ 218,631.91","R$ 63,960.39",38,29.25%,"R$ 5,753.47",11.38%
4,PE,"R$ 208,457.44","R$ 58,183.95",55,27.91%,"R$ 3,790.14",10.85%
3,MG,"R$ 207,614.69","R$ 61,122.25",44,29.44%,"R$ 4,718.52",10.80%
8,SC,"R$ 198,143.47","R$ 60,272.32",40,30.42%,"R$ 4,953.59",10.31%
7,RS,"R$ 187,048.31","R$ 49,795.96",43,26.62%,"R$ 4,349.96",9.73%
1,ES,"R$ 181,113.70","R$ 49,838.07",39,27.52%,"R$ 4,643.94",9.43%
5,PR,"R$ 174,136.19","R$ 46,155.64",35,26.51%,"R$ 4,975.32",9.06%
0,BA,"R$ 157,082.99","R$ 42,117.30",42,26.81%,"R$ 3,740.07",8.18%
2,GO,"R$ 142,428.02","R$ 42,028.11",36,29.51%,"R$ 3,956.33",7.41%


## 10. O motor da análise: agregações (groupby)

Nos slides, a lógica central é:
- `receita -> sum`
- `margem -> mean` ou cálculo derivado
- `produto -> count`
fileciteturn5file0

### Tarefa
Escolha **duas dimensões diferentes** e construa tabelas agregadas mostrando que você entendeu essa lógica.

Exemplos:
- por canal
- por categoria
- por UF
- por mês


## 11. Campeões e vilões (Top N)

A aula destaca o poder dos rankings curtos e focados. fileciteturn5file0

### Tarefa
Crie:
- Top 10 produtos por **lucro**
- Top 10 produtos por **receita**
- 5 categorias ou UFs com **pior margem** ou pior desempenho

### Perguntas
1. Quem são os “campeões”?

"Notebook pro", "Smartphone X" e "Tablet Plus"

2. Quem são os “vilões”?

"telefonia" e "informatica"

3. Que ação tática poderia ser tomada com base nisso?

Para os campeões é otimizar campanhas de marketing para os produtos visando a estabilidade do sucesso de vendas.
Para os vilões é investigar as baixas margens como altos custos de aquisição, forte concorrência e precificação inadequada

In [22]:
# Top 10 produtos por lucro
top_10_produtos_lucro = df.groupby('produto')['lucro'].sum().nlargest(10).reset_index()
display(top_10_produtos_lucro.rename(columns={'lucro': 'lucro_total'}))

# Top 10 produtos por receita
top_10_produtos_receita = df.groupby('produto')['receita'].sum().nlargest(10).reset_index()
display(top_10_produtos_receita.rename(columns={'receita': 'receita_total'}))

# 5 categorias com pior margem
piores_5_categorias = analise_categoria.sort_values(by='margem_calculada', ascending=True).head(5)
display(piores_5_categorias)

# 5 UFs com pior margem
piores_5_ufs = analise_uf.sort_values(by='margem_calculada', ascending=True).head(5)
display(piores_5_ufs)

,produto,lucro_total
0,Notebook Pro,153094.84
1,Smartphone X,118198.25
2,Tablet Plus,101777.33
3,Monitor 27,65995.31
4,Teclado Mecânico,33939.07
5,Caixa de Som,28316.37
6,Headset Pro,26689.65
7,Mouse Gamer,16485.33


,produto,receita_total
0,Notebook Pro,594132.74
1,Smartphone X,489572.08
2,Tablet Plus,376720.37
3,Monitor 27,211243.70
4,Teclado Mecânico,79973.75
5,Caixa de Som,68956.76
6,Headset Pro,63943.49
7,Mouse Gamer,36949.33


,categoria,receita_total,lucro_total,quantidade_pedidos,margem_calculada,ticket_medio
2,Telefonia,"R$ 866,292.45","R$ 219,975.58",109,25.39%,"R$ 7,947.64"
0,Informática,"R$ 805,376.44","R$ 219,090.15",93,27.20%,"R$ 8,659.96"
3,Áudio,"R$ 132,900.25","R$ 55,006.02",111,41.39%,"R$ 1,197.30"
1,Periféricos,"R$ 116,923.08","R$ 50,424.40",107,43.13%,"R$ 1,092.74"


,uf,receita_total,lucro_total,quantidade_pedidos,margem_calculada,ticket_medio,participacao_receita_percentual
5,PR,"R$ 174,136.19","R$ 46,155.64",35,26.51%,"R$ 4,975.32",9.06%
7,RS,"R$ 187,048.31","R$ 49,795.96",43,26.62%,"R$ 4,349.96",9.73%
0,BA,"R$ 157,082.99","R$ 42,117.30",42,26.81%,"R$ 3,740.07",8.18%
1,ES,"R$ 181,113.70","R$ 49,838.07",39,27.52%,"R$ 4,643.94",9.43%
4,PE,"R$ 208,457.44","R$ 58,183.95",55,27.91%,"R$ 3,790.14",10.85%


## 12. Análise temporal

A aula reforça que sem tempo a análise fica estática. fileciteturn5file0

### Tarefa
1. Converta `data_venda` para data, se necessário
2. Crie colunas auxiliares como:
   - mês
   - trimestre (opcional)
3. Gere uma tabela agregada por mês com:
   - receita
   - lucro
   - ticket médio ou margem

### Depois responda
1. O lucro parece crescer ao longo do tempo?

O lucro apresenta flutuações ao longo do ano. Observamos que os meses de julho (R$ 57.913,44) e dezembro (R$$ 57.913,44) e dezembro (R$ 62.442,65) mostram os maiores lucros, indicando um desempenho mais forte no segundo semestre, especialmente no final do ano, mas sem um crescimento linear e constante a cada mês.

2. O último trimestre foi melhor que o anterior?

O último trimestre (outubro, novembro, dezembro) totalizou um lucro de R$ 154.398,02. O trimestre anterior (julho, agosto, setembro) totalizou um lucro de R$ 154.398,02. O trimestre anterior (julho, agosto, setembro) totalizou um lucro de R$ 157.550,54. Portanto, o último trimestre foi ligeiramente inferior ao anterior em termos de lucro total.

3. Quais meses concentram as vendas de fim de ano?

Claramente, Novembro e Dezembro são os meses que concentram as vendas de fim de ano, com Dezembro apresentando a maior receita (R$ 200.909,84) e o maior lucro (R$$ 200.909,84) e o maior lucro (R$ 62.442,65) de todo o ano.




In [ ]:
df['data_venda'] = pd.to_datetime(df['data_venda'])
df['mes'] = df['data_venda'].dt.to_period('M')

analise_temporal_mensal = df.groupby('mes').agg(
    receita_total=('receita', 'sum'),
    lucro_total=('lucro', 'sum'),
    margem_media=('margem_lucro', 'mean'),
    # Calculando o ticket médio mensal
    ticket_medio_mensal=('receita', lambda x: x.sum() / df.loc[x.index, 'pedido_id'].nunique())
).reset_index()

# Formatar para melhor visualização
analise_temporal_mensal['receita_total'] = analise_temporal_mensal['receita_total'].map('R$ {:,.2f}'.format)
analise_temporal_mensal['lucro_total'] = analise_temporal_mensal['lucro_total'].map('R$ {:,.2f}'.format)
analise_temporal_mensal['margem_media'] = analise_temporal_mensal['margem_media'].map('{:,.2%}'.format)
analise_temporal_mensal['ticket_medio_mensal'] = analise_temporal_mensal['ticket_medio_mensal'].map('R$ {:,.2f}'.format)

display(analise_temporal_mensal.sort_values(by='mes'))

## 13. Fluxo de investigação do analista

A aula apresenta a sequência:
**Pergunta -> Colunas necessárias -> Tipo de resposta/gráfico**. fileciteturn5file0

### Tarefa
Preencha, em markdown, pelo menos 3 exemplos no formato:

- **Pergunta de negócio:** ...
- **Colunas necessárias:** ...
- **Métrica/KPI:** ...
- **Tipo de resposta esperada:** tabela, ranking, tendência etc.

**Pergunta de negócio:** Como a receita e o lucro evoluem ao longo do tempo, mensalmente?
- **Colunas necessárias:** `data_venda`, `receita`, `lucro`, `pedido_id`
- **Métrica/KPI:** Receita Total Mensal, Lucro Total Mensal, Margem Média Mensal, Ticket Médio Mensal
- **Tipo de resposta esperada:** Tabela de tendência e, opcionalmente, gráficos de linha para visualização temporal.


## 14. O insight exige palavras

Segundo a aula, o mercado valoriza quem comunica decisões, não apenas quem roda código. fileciteturn5file0

### Tarefa
Escolha **duas tabelas agregadas** que você criou e escreva, para cada uma, **1 ou 2 frases de interpretação executiva**.

Exemplo de estilo:
- “O canal X concentra maior volume de receita, mas o canal Y apresenta margem superior.”
- “Recomenda-se revisar custos/expansão/priorização com base nesse resultado.”

---
**Interpretação para Análise por Canal:**
O canal Marketplace demonstra o maior volume de receita e lucro, consolidando sua posição como principal motor de vendas. Contudo, o canal Online exibe uma margem de lucro superior, indicando maior eficiência e potencial para otimização de rentabilidade se suas estratégias forem aplicadas a outros canais.

---
**Interpretação para Análise por Categoria:**
A categoria Telefonia lidera em receita, mas sua margem de lucro é a mais baixa, sugerindo que um alto volume de vendas não se traduz automaticamente em alta rentabilidade. Por outro lado, a categoria Periféricos, apesar de um volume menor, apresenta a maior margem de lucro, indicando uma excelente eficiência e possivelmente um posicionamento de mercado mais vantajoso


## 15. Case Varejo Brasil — Entrega final

A aula propõe quatro entregas centrais para o gestor. fileciteturn5file0

### Sua missão final
Organize o notebook para entregar:

1. **KPIs Globais**
2. **Tabelas por Dimensões Críticas**
   - por canal
   - por categoria
3. **Top 10 Produtos por Lucro**
4. **Conclusão final com pelo menos uma decisão tática**

### Importante
A conclusão final deve ser escrita em linguagem clara, direta e orientada à decisão.


Conclusão

1. KPIs Globais da Operação:

Receita Total: R 1.921.492,22−∗∗LucroTotal:∗∗R  544.496,15
Margem de Lucro: 28,34%
Ticket Médio: R$ 4.574,98
Estes indicadores mostram uma operação com volume significativo, boa rentabilidade e um ticket médio que sugere vendas de itens de valor considerável.

2. Análise por Dimensões Críticas:

Por Canal de Venda:

O Marketplace é o principal motor de receita e lucro, demonstrando grande volume.
O canal Online, apesar de menor volume, possui a maior margem de lucro (29,56%), indicando maior eficiência e potencial para replicar suas estratégias de rentabilidade em outros canais.
O canal Loja Física tem desempenho sólido, mas com margem ligeiramente inferior ao Online.
Por Categoria de Produto:

Telefonia e Informática lideram em receita, mas a Telefonia apresenta a menor margem (25,39%), sugerindo que o alto volume não se traduz em alta rentabilidade. Recomenda-se investigar custos e precificação nesta categoria.
Periféricos e Áudio mostram as maiores margens (43,13% e 41,39% respectivamente), indicando alta eficiência e bom posicionamento de mercado, mesmo com volumes menores.
3. Top 10 Produtos por Lucro: Os produtos "Notebook Pro", "Smartphone X" e "Tablet Plus" são os maiores contribuidores para o lucro, sendo os "campeões" da operação. Eles devem ter sua disponibilidade e esforços de marketing priorizados.

4. Decisão Tática Recomendada:

Com base na análise, é fundamental rever a estratégia de precificação e custos da categoria 'Telefonia'. Embora gere alta receita, sua baixa margem de lucro impacta a rentabilidade geral. Paralelamente, explorar as práticas de otimização de custos e precificação do canal Online e das categorias de alta margem como "Periféricos" para identificar oportunidades de aplicar essas eficiências em toda a operação, especialmente no canal Marketplace para maximizar o lucro de seu alto volume.

## 16. Desafio extra (opcional)

Monte uma pequena tabela executiva final com:
- Receita Total
- Lucro Total
- Margem
- Ticket Médio

e depois crie uma segunda tabela por canal com os mesmos indicadores.

Compare os resultados e explique:
- o KPI global esconde alguma diferença importante entre os canais?


In [ ]:
# Desafio extra opcional


## 17. Entrega esperada

Seu notebook deve demonstrar:
- organização
- fórmulas corretas de KPI
- uso adequado de `groupby`, `agg`, `sort_values`
- leitura temporal coerente
- interpretação escrita dos resultados

### Mensagem principal da aula
Transformar linhas de dados limpos em **apoio real para decisões de negócio**. fileciteturn5file0
